server/service/orbit_client.go

```diff
@@ -68,26 +127,115 @@ func (oc *OrbitClient) DoEnroll() (string, error) {
+	switch {
+	case err == nil:
+		return string(orbitNodeKey), nil
+	case errors.Is(err, fs.ErrNotExist):
+		// OK, if there's no orbit node key, proceed to enroll.
+	default:
+		return "", fmt.Errorf("read orbit node key file: %w", err)
 	}
+	var (
+		orbitNodeKey_        string
+		endpointDoesNotExist bool
+	)
+	if err := retry.Do(
+		func() error {
+			var err error
+			orbitNodeKey_, err = oc.enrollAndWriteNodeKeyFile()
+			switch {
+			case err == nil:
+				return nil
+			case errors.Is(err, notFoundErr{}):
+				// Do not retry if the endpoint does not exist.
+				endpointDoesNotExist = true
+				return nil
+			default:
+				log.Info().Err(err).Msg("enroll failed, retrying")
+				return err
+			}
+		},
+		retry.WithInterval(constant.OrbitEnrollRetrySleep),
+		retry.WithMaxAttempts(constant.OrbitEnrollMaxRetries),
+	); err != nil {
+		return "", fmt.Errorf("orbit node key enroll failed, attempts=%d", constant.OrbitEnrollMaxRetries)
+	}
+	if endpointDoesNotExist {
+		return "", errors.New("enroll endpoint does not exist")
+	}
+	return orbitNodeKey_, nil
+}
 
-	return resp.Flags, nil
+func (oc *OrbitClient) enrollAndWriteNodeKeyFile() (string, error) {
+	orbitNodeKey, err := oc.enroll()
+	if err != nil {
+		return "", fmt.Errorf("enroll request: %w", err)
+	}
+	if err := os.WriteFile(oc.nodeKeyFilePath, []byte(orbitNodeKey), constant.DefaultFileMode); err != nil {
+		return "", fmt.Errorf("write orbit node key file: %w", err)
+	}
+	return orbitNodeKey, nil
 }
 
-func (oc *OrbitClient) Ping() error {
-	verb, path := "HEAD", "/api/fleet/orbit/ping"
-	err := oc.request(verb, path, nil, nil)
+func (oc *OrbitClient) authenticatedRequest(verb string, path string, params interface{}, resp interface{}) error {
+	nodeKey, err := oc.getNodeKeyOrEnroll()
+	if err != nil {
+		return err
+	}
 
-	if err == nil || errors.Is(err, notFoundErr{}) {
-		// notFound is ok, it means an old server without the capabilities header
+	s := params.(setOrbitNodeKeyer)
+	s.setOrbitNodeKey(nodeKey)
+
+	switch oc.request(verb, path, params, resp); {
+	case err == nil:
+		oc.setEnrolled(true)
 		return nil
+	case errors.Is(err, ErrUnauthenticated):
+		if err := os.Remove(oc.nodeKeyFilePath); err != nil {
+			log.Info().Err(err).Msg("remove orbit node key")
+		}
+		oc.setEnrolled(false)
+		return err
+	default:
+		return err
 	}
+}
 
-	return err
+func (oc *OrbitClient) Enrolled() bool {
+	oc.enrolledMu.Lock()
+	defer oc.enrolledMu.Unlock()
+
+	return oc.enrolled
+}
+
+func (oc *OrbitClient) setEnrolled(v bool) {
+	oc.enrolledMu.Lock()
+	defer oc.enrolledMu.Unlock()
+
+	oc.enrolled = v
+}
+
+func (oc *OrbitClient) LastRequestError() error {
+	oc.lastReqErrMu.Lock()
+	defer oc.lastReqErrMu.Unlock()
+
+	return oc.lastReqErr
+}
+
+func (oc *OrbitClient) setLastRequestError(err error) {
+	oc.lastReqErrMu.Lock()
+	defer oc.lastReqErrMu.Unlock()
+
+	oc.lastReqErr = fmt.Errorf("%s: %w", time.Now().UTC().Format("2006-01-02T15:04:05Z"), err)
```

Comment:
do you remember if `err` includes the request path/method? I'm thinking it would be nice to have it to know which request failed.

Reply:
![Screen Shot 2022-10-03 at 14 56 13](https://user-images.githubusercontent.com/2073526/193645847-830e5ea0-2625-4988-877b-31cf8a1c545b.png)